src/libs/actions/Task.js

```diff
@@ -0,0 +1,102 @@
+import * as Report from './Report';
+
+/**
+ * Assign a task to a user
+ * @param {String} parentReportActionID
+ * @param {String} parentReportID
+ * @param {String} taskReportID
+ * @param {String} name
+ * @param {String} description
+ * @param {String} assignee
+ * @param {String} ownerEmail
+ *
+ */
+
+function assignTask(parentReportActionID, parentReportID, taskReportID, name, description, assignee, assigneeChatReportID, ownerEmail) {
+    let parentReport;
+
+    // If there isn't a parentReportId provided and assignee is given, we need to optimistically create a new parent chat report
+    // between the task creator and the assignee.
+    if (!parentReportID && assignee) {
+        parentReport = ReportUtils.getChatByParticipants([assignee]);
+    }
+
+    if (!parentReportID && !parentReport && assignee) {
+        parentReport = ReportUtils.buildOptimisticChatReport([assignee]);
+    }
+
+    // If parentReport is defined, use its reportID, otherwise, use the provided parentReportID
+    const finalParentReportID = parentReport ? parentReport.reportID : parentReportID;
+
+    // Open the parent report if assignee is given
+    if (assignee) {
+        Report.openReport(finalParentReportID, [assignee]);
+    }
+
+    const optimisticTaskReport = ReportUtils.buildOptimisticTaskReport(ownerEmail, assignee, finalParentReportID, parentReportActionID, name, description);
+    const optimisticCreatedAction = ReportUtils.buildOptimisticCreatedReportAction(ownerEmail);
+
+    // AddCommentReportAction on the parent chat report
+    const optimisticAddCommentReportAction = ReportUtils.buildOptimisticAddCommentReportAction(finalParentReportID, optimisticTaskReport.reportID);
+
+    const optimisticData = [
+        {
+            method: CONST.ONYX.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${finalParentReportID}`,
+            value: optimisticCreatedAction,
+        },
+        {
+            method: CONST.ONYX.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${finalParentReportID}`,
+            value: parentReport,
+        },
+        {
+            method: CONST.ONYX.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${optimisticTaskReport.reportID}`,
+            value: optimisticTaskReport,
+        },
+        optimisticAddCommentReportAction,
+    ];
+
+    const successData = {};
+
+    const failureData = {
+        undoActions: [
+            {
+                method: 'DELETE_TASK_REPORT',
+                reportID: optimisticTaskReport.reportID,
+            },
+            {
+                method: 'DELETE_REPORT_ACTION',
+                reportID: finalParentReportID,
+                actionID: optimisticCreatedAction.actionID,
+            },
+            {
+                method: 'DELETE_ADD_COMMENT_REPORT_ACTION',
+                reportID: finalParentReportID,
+                actionID: optimisticAddCommentReportAction.actionID,
+            },
+        ],
+    };
```

Comment:
for failure data, we can just clear the keys that were set in the optimistic data.

For example, one of the optimistic items is

{
            method: CONST.ONYX.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${finalParentReportID}`,
            value: optimisticCreatedAction,
        },

So for failure you can do

{
	method: CONST.ONYX.METHOD.SET,
	key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${finalParentReportID}`,
	value: null
}